### PR TITLE
Add email link sign in support

### PR DESCRIPTION
**Pull request by renkelvin**

### CODE REVIEW

File: Auth/FirebaseAuthUI/FUIAuthBaseViewController_Internal.h [hunk 1]
```diff
@@ -38,57 +38,51 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)showAlertWithMessage:(NSString *)message;
 
+/** @fn showAlertWithMessage:
+    @brief Displays an alert view with given title and message on top of the current view
+    controller.
```
Comment (Yue-Wang-Google) [inline Auth/FirebaseAuthUI/FUIAuthBaseViewController_Internal.h]: indent

File: EmailAuth/FirebaseEmailAuthUI/FUIConfirmEmailViewController.m [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+- (instancetype)initWithAuthUI:(FUIAuth *)authUI {
+  return [self initWithNibName:NSStringFromClass([self class])
+                        bundle:[FUIAuthUtils bundleNamed:FUIEmailAuthBundleName]
+                        authUI:authUI];
+}
+
+- (instancetype)initWithNibName:(NSString *)nibNameOrNil
+                         bundle:(NSBundle *)nibBundleOrNil
+                         authUI:(FUIAuth *)authUI {
+
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIConfirmEmailViewController.m]: remove new line

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 3]
```diff
@@ -190,10 +203,243 @@ - (void)signInWithDefaultValue:(nullable NSString *)defaultValue
...
+
+  // Retrieve url parameters from continueUrl
+  NSMutableDictionary *urlParameterDict= [NSMutableDictionary dictionary];
+  NSURLComponents *continueURLComponents = [NSURLComponents componentsWithString:continueURLString];
+  for (NSURLQueryItem *queryItem in continueURLComponents.queryItems) {
+    urlParameterDict[queryItem.name] = queryItem.value;
+  }
+  // Retrieve parameters from local storage
+  NSMutableDictionary *localParameterDict = [NSMutableDictionary dictionary];
+  localParameterDict[kEmailLinkSignInEmailKey] = [GULUserDefaults.standardUserDefaults stringForKey:kEmailLinkSignInEmailKey];
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: 100 char limit

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 4]
```diff
@@ -190,10 +203,243 @@ - (void)signInWithDefaultValue:(nullable NSString *)defaultValue
...
+  localParameterDict[@"ui_sid"] = [GULUserDefaults.standardUserDefaults stringForKey:@"ui_sid"];
+
+  // Handling flows
+  NSString *urlSessionID = urlParameterDict[@"ui_sid"];
+  NSString *localSessionID = localParameterDict[@"ui_sid"];
+  BOOL sameDevice = urlSessionID && localSessionID && [urlSessionID isEqualToString:localSessionID];
+
+  if (sameDevice) { // Same device
+    if (urlParameterDict[@"ui_pid"] != nil) { // Unverified provider linking
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: you dont need ! = nil. it's verbose

Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: conditional comment usually goes into the next line. also full stop in the end.  same below

if (foo) {
  // process foo.
  processFoo();
}

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 5]
```diff
@@ -190,10 +203,243 @@ - (void)signInWithDefaultValue:(nullable NSString *)defaultValue
...
+    }
+  }
+
+  return YES;
+}
+
+- (void)handleUnverifiedProviderLinking:(NSString *)providerID
+                                  email:(NSString *)email {
+  if ([providerID isEqualToString:FIRFacebookAuthProviderID]) {
+    NSData *unverifiedProviderCredentialData = [GULUserDefaults.standardUserDefaults objectForKey:kEmailLinkSignInLinkingCredentialKey];
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: 100 char limit

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 6]
```diff
@@ -400,6 +647,56 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
...
+          @"You already have an account\n \n You've already used %@. You "
+          "can connect your %@ account with %@ by signing in with Email "
+          "link below. \n \n For this flow to successfully connect your "
+          "account with this email, you have to open the link on the same "
+          "device or browser.", email, providerName, email];
+      [FUIAuthBaseViewController
+          showAlertWithTitle:@"Sign in"
+                     message:message
+                 actionTitle:@"Sign in"
+               actionHandler:^{
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: make this block a temp variable so the call can be more readable

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 7]
```diff
@@ -467,6 +764,51 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
 
 #pragma mark - Private
 
+- (void)generateURLParametersAndLocalCache:(NSString *)email linkingProvider:(NSString *)linkingProvider {
+  NSURL *url = self.actionCodeSettings.URL;
+  NSURLComponents *urlComponents = [NSURLComponents componentsWithString:url.absoluteString];
+  NSMutableArray<NSURLQueryItem *> *urlQuertItems = [NSMutableArray array];
+
+  // kEmailLinkSignInEmailKey
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: remove comment.

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 8]
```diff
@@ -467,6 +764,51 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
...
+  [GULUserDefaults.standardUserDefaults setObject:email forKey:kEmailLinkSignInEmailKey];
+
+  // ui_auid
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: remove comment.

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 9]
```diff
@@ -467,6 +764,51 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
...
+  if (self.authUI.auth.currentUser.isAnonymous && self.authUI.shouldAutoUpgradeAnonymousUsers) {
+    NSString *auid = self.authUI.auth.currentUser.uid;
+
+    NSURLQueryItem *anonymousUserIDQueryItem =
+        [NSURLQueryItem queryItemWithName:@"ui_auid" value:auid];
+    [urlQuertItems addObject:anonymousUserIDQueryItem];
+  }
+
+  // ui_sid
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: remove comment.

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 10]
```diff
@@ -467,6 +764,51 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
...
+  NSInteger ui_sid = arc4random_uniform(999999999);
+  NSString *sidString = [NSString stringWithFormat:@"%ld", (long)ui_sid];
+  [GULUserDefaults.standardUserDefaults setObject:sidString forKey:@"ui_sid"];
+
+  NSURLQueryItem *sessionIDQueryItem =
+      [NSURLQueryItem queryItemWithName:@"ui_sid" value:sidString];
+  [urlQuertItems addObject:sessionIDQueryItem];
+
+  // ui_sd
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: remove comment

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m [hunk 11]
```diff
@@ -467,6 +764,51 @@ - (void)handleAccountLinkingForEmail:(NSString *)email
...
+  NSString *sameDeviceValueString;
+  if (self.forceSameDevice) {
+    sameDeviceValueString = @"1";
+  } else {
+    sameDeviceValueString = @"0";
+  }
+  NSURLQueryItem *sameDeviceQueryItem = [NSURLQueryItem queryItemWithName:@"ui_sd" value:sameDeviceValueString];
+  [urlQuertItems addObject:sameDeviceQueryItem];
+
+  // ui_pid
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m]: remove comment.

File: EmailAuth/FirebaseEmailAuthUI/FUIEmailEntryViewController.m [hunk 12]
```diff
@@ -196,17 +211,65 @@ - (void)onNext:(NSString *)emailText {
...
+                                                 [self.navigationController popToRootViewControllerAnimated:YES];
+                                               }
+                                     presentingViewController:self];
+                                      }
+                                       dismissTitle:@"Back"
+                                     dismissHandler:^{
+                                       [self.navigationController popToRootViewControllerAnimated:YES];
+                                     }
+                           presentingViewController:self
+       ];
```
Comment (Yue-Wang-Google) [inline EmailAuth/FirebaseEmailAuthUI/FUIEmailEntryViewController.m]: ] goes to the previous line.

File: samples/objc/FirebaseUI-demo-objc/FUIAppDelegate.m [hunk 13]
```diff
@@ -47,6 +47,24 @@ - (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceAppl
   return [self handleOpenUrl:url sourceApplication:sourceApplication];
 }
 
+- (BOOL)application:(UIApplication *)application continueUserActivity:(nonnull NSUserActivity *)userActivity restorationHandler:
```
Comment (Yue-Wang-Google) [inline samples/objc/FirebaseUI-demo-objc/FUIAppDelegate.m]: 100 char limit

File: samples/objc/FirebaseUI-demo-objc/FUIAppDelegate.m [hunk 14]
```diff
@@ -47,6 +47,24 @@ - (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceAppl
...
+#if defined(__IPHONE_12_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_12_0)
+  (nonnull void (^)(NSArray<id<UIUserActivityRestoring>> *_Nullable))restorationHandler {
+#else
+  (nonnull void (^)(NSArray *_Nullable))restorationHandler {
+#endif  // __IPHONE_12_0
+  BOOL handled = [[FIRDynamicLinks dynamicLinks] handleUniversalLink:userActivity.webpageURL
+                                                          completion:^(FIRDynamicLink * _Nullable dynamicLink,
```
Comment (Yue-Wang-Google) [inline samples/objc/FirebaseUI-demo-objc/FUIAppDelegate.m]: make a temp variable so you have better indentation.

File: samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m [hunk 15]
```diff
@@ -344,31 +354,40 @@ - (void)showAlertWithTitlte:(NSString *)title message:(NSString *)message {
...
   for (NSIndexPath *indexPath in selectedRows) {
     if (indexPath.section == kSectionsProviders) {
       id<FUIAuthProvider> provider;
       switch (indexPath.row) {
         case kIDPEmail:
-          provider = [[FUIEmailAuth alloc] init];
+          if (useEmaiLink) {
+            // ActionCodeSettings for email link sign-in.
+            FIRActionCodeSettings *actionCodeSettings = [[FIRActionCodeSettings alloc] init];
+            [actionCodeSettings setURL:[NSURL URLWithString:@"https://fb-sa-1211.appspot.com"]];
```
Comment (Yue-Wang-Google) [inline samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m]: this is  property of FIRActionCodeSettings. no need to use setURL. use property syntax.

File: samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m [hunk 16]
```diff
@@ -344,31 +354,40 @@ - (void)showAlertWithTitlte:(NSString *)title message:(NSString *)message {
...
+            actionCodeSettings.handleCodeInApp = YES;
+            [actionCodeSettings setIOSBundleID:[[NSBundle mainBundle] bundleIdentifier]];
```
Comment (Yue-Wang-Google) [inline samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m]: having setIOSBundleID API is pretty stupid. probably we need to remove that in the ios-sdk and make iOSBundleID property readwrite.

Comment (Yue-Wang-Google) [reply samples/objc/FirebaseUI-demo-objc/Samples/Auth/FUIAuthViewController.m]: No action needed this time.

**Comment by morganchen12** [comment]
https://github.com/firebase/FirebaseUI-iOS/pull/593 was merged, so you can write tests for this CL in the `UITests/` folder, but you can also do that in a separate change.

**Pull request merged**